src/charm.py

```diff
@@ -100,23 +110,86 @@ def _on_config_changed(self, event):
 
     def publish_config(self):
         """Generate config file and publish to all workers."""
-        mimir_config = self.coordinator.build_config(dict(self.config))
+        mimir_config = self.coordinator.build_config(self._s3_storage)
         self.cluster_provider.publish_configs(mimir_config)
 
+    def create_minio_buckets(self, conn: dict, bucket_names: list):
+        """Create Minio buckets."""
+        try:
+            client = Minio(
+                endpoint=f'{conn["service"]}.{conn["namespace"]}.svc.cluster.local:{conn["port"]}',
+                access_key=conn["access_key"],
+                secret_key=conn["secret_key"],
+                secure=conn["secure"],
+            )
+            for bucket in bucket_names:
+                found = client.bucket_exists(bucket)
+                if not found:
+                    client.make_bucket(bucket)
+                else:
+                    logger.info("Bucket already exists")
+        except S3Error:
+            logger.error("Error creating S3 buckets")
+
+    def _parse_s3_data(self, s3_data):
+        data_passed_dict = dict(item.split(": ") for item in s3_data.split("\n") if item)
+        data_passed_dict["secure"] = data_passed_dict.get("secure", "").lower() == "true"
+
+        return {
```

Comment:
Where is this schema coming from? Should we have a dataclass or pydantic model for it?

Reply:
Multiline text coming through from [Charmed Minio](https://github.com/canonical/minio-operator/blob/main/src/charm.py#L161-L172), removed now after moving to s3-intgerator